ggml-hexagon: create generalized functions for cpu side op

chraac opened the pull request:
This PR refactors the Hexagon backend to generalize CPU-side operations and improve code maintainability. The primary goal is to reduce code duplication by merging similar operation handlers and standardizing buffer management.

### Key Changes

1.  **Generalized Operation Handlers**:
    *   Merged `ggml_hexagon_mul_mat` into `ggml_hexagon_binary`.
    *   Merged `ggml_hexagon_mul_mat_id` into `ggml_hexagon_add_id` (renamed to `ggml_hexagon_binary_id`).
    *   These functions are now templated with `<bool _IsSrc0Constant>` to efficiently handle cases where the first operand (like weights) is constant, optimizing cache maintenance.

2.  **Refactored Buffer Management**:
    *   Introduced `enum dsp_buffer_type` to explicitly define buffer roles:
        *   `DSP_BUFFER_TYPE_DSP_WRITE_CPU_READ`
        *   `DSP_BUFFER_TYPE_CPU_WRITE_DSP_READ`
        *   `DSP_BUFFER_TYPE_CONSTANT`

### Test

[test-backend-ops-all.release.htp.4c394c250.log](https://drive.google.com/file/d/1jL3dt93gKDtpKqfZpt2lSo3D0xfntwUb/view?usp=sharing)

chraac committed:
refactor: replace ggml_hexagon_mul_mat with template-based binary operation for improved flexibility

refactor: replace ggml_hexagon_mul_mat_id with template-based binary operation for improved flexibility

refactor: initialize buffer types and streamline dspqueue_buffers_init calls for clarity

refactor: remove redundant buffer checks in hexagon supported operations

add missing include to fix weak symbol warning

add ggml_hexagon_op_generic

refactor: simplify tensor operation initialization and buffer management in hexagon implementation

refactor: streamline hexagon operation initialization and buffer management

refactor: update function signatures and streamline request handling in hexagon operations

ggml-hexagon: clean up code formatting and improve unary operation handling

fix: add support for permuted F16 tensors and enhance quantization checks in matrix operations

# Conflicts:
#	ggml/src/ggml-hexagon/ggml-hexagon.cpp

chraac committed:
refactor: replace ggml_hexagon_mul_mat with template-based binary operation for improved flexibility

chraac committed:
refactor: replace ggml_hexagon_mul_mat_id with template-based binary operation for improved flexibility

chraac committed:
refactor: initialize buffer types and streamline dspqueue_buffers_init calls for clarity

chraac committed:
add comment

chraac committed:
refactor: remove redundant buffer checks in hexagon supported operations

chraac committed:
wip

chraac committed:
add missing include to fix weak symbol warning

chraac commented on ggml/src/ggml-hexagon/ggml-hexagon.cpp:
```diff
@@ -2560,7 +2384,7 @@ static void ggml_hexagon_binary(const struct ggml_tensor * op, uint32_t flags) {
 }
 
-static void ggml_hexagon_add_id(const struct ggml_tensor * op, uint32_t flags) {
+template <bool _IsSrc0Constant> static void ggml_hexagon_binary_id(const struct ggml_tensor * op, uint32_t flags) {
```
We can probably refactor these op functions into a single ggml_hexagon_generic function that handles null sources. This would reduce code duplication significantly. WDYT? @max-krasnyansky

chraac committed:
add ggml_hexagon_op_generic

chraac committed:
refactor: simplify tensor operation initialization and buffer management in hexagon implementation

chraac committed:
refactor: streamline hexagon operation initialization and buffer management

chraac committed:
refactor: update function signatures and streamline request handling in hexagon operations

chraac committed:
wip

chraac committed:
Merge branch 'master' into dev-hexagon-refactoring-part2

chraac committed:
Merge tag 'b7207' into dev-hexagon-refactoring-part2

chraac committed:
Merge tag 'b7268' into dev-hexagon-refactoring-part2

chraac committed:
Merge branch 'master' into dev-hexagon-refactoring-part2

chraac committed:
Merge tag 'b7345' into dev-hexagon-refactoring-part2

chraac committed:
Merge branch 'master' into dev-hexagon-refactoring-part2

chraac committed:
Merge branch 'master' into dev-hexagon-refactoring-part2

max-krasnyansky commented:
@chraac 
This looks good overall. Nice to see a good chunk of the code getting more generic / more reuse.
Could you please do a clean rebase / squash, mark it as read and will merge (or iterate a bit if needed). 
Also let's not add CamelCase `IsSrc0Static` just use the underscores.  

btw We can optimize things here a bit further here.  
We can skip NPU (DSP) cache flush for all tensors that are outputs from previous Ops. 
i.e. src0_is_static would become `is_static || is_output_from_htp`.  
No need to complicate things in this PR though we can do that as a follow up.

chraac committed:
Merge tag 'b7438' into dev-hexagon-refactoring-part2

chraac committed:
Merge branch 'master' into dev-hexagon-refactoring-part2

# Conflicts:
#	ggml/src/ggml-hexagon/ggml-hexagon.cpp

chraac committed:
ggml-hexagon: clean up code formatting and improve unary operation handling

chraac committed:
wip

chraac committed:
rename

chraac committed:
fix: add support for permuted F16 tensors and enhance quantization checks in matrix operations

chraac committed:
Merge branch 'master' into dev-hexagon-refactoring-part2

chraac committed:
Merge branch 'master' into dev-hexagon-refactoring-part2

max-krasnyansky committed:
hexagon: fix merge conflicts

max-krasnyansky committed:
hexagon: minor cleanup for buffer support checks

max-krasnyansky committed:
hexagon: factor out op_desc and the overal op logging

max-krasnyansky committed:
hexagon: further simplify and cleanup op dispatch logic

max-krasnyansky committed:
snapdragon: update adb scripts to use llama-cli and llama-completion

chraac committed:
Merge branch 'master' into dev-hexagon-refactoring-part2

# Conflicts:
#	ggml/src/ggml-hexagon/ggml-hexagon.cpp

chraac commented:
> @chraac
> 
> Sorry for the delay. I finally had the chance to play with this. Ended up doing a bunch of updates on top and lots of testing. It's shaping up really nice with lots of redundant code gone. I think most of those original comments about cache management are no longer helpful.
> 
> Take a look at this branch. Please pull into this PR and test your scenarios. https://github.com/qualcomm/llama.cpp/tree/hexagon-refactor-op-dispatch
> 
> btw don't try to git merge though. I already rebased and squashed your commits and they are all included.

Nice work! I've pulled the changes into my branch and verified that it works well with my test cases.
Besides, did notice that `vec_dot_f16_f32` now falls back to the non-SIMD implementation to handle leftover elements. Planning to look into that later to see if I can fix the SIMD implementation.

max-krasnyansky commented:
> @max-krasnyansky Nice work! I've pulled the changes into my branch and verified that it works well with my test cases. Besides, did notice that `vec_dot_f16_f32` now falls back to the non-SIMD implementation to handle leftover elements. Planning to look into that later to see if I can fix the SIMD implementation.

Awesome. Thanks. 
And yes, fp16 matmul definitely needs more updates. 

@chraac looks like I missed EditorConfig check failure. A couple of lines have extra spaces.
```
ggml/src/ggml-hexagon/ggml-hexagon.cpp:
	2173: Trailing whitespace
	2324: Trailing whitespace
```

When you get the chance can you please fix it so that we can pass CI and I'll merge.

chraac committed:
fix pipeline failure
